fortanix-sgx-tools/src/bin/ftxsgx-elf2sgxs.rs

```diff
@@ -154,11 +166,62 @@ macro_rules! check_size {
             );
         }
     }};
+    ($name:ident == $size:expr) => {{
+        let size = $name.size();
+        if size != $size {
+            bail!(
+                "Dynamic symbol {} doesn't have the right size. Expected size {}, got {}.",
+                stringify!($name),
+                $size,
+                size
+            );
+        }
+    }};
 }
 
 impl<'a> LayoutInfo<'a> {
+
+    // Check version defined in rust-lang assembly code is supported, see .note.x86_64-fortanix-unknown-sgx section in https://github.com/rust-lang/rust/blob/master/src/libstd/sys/sgx/abi/entry.S
+    fn check_toolchain_version(elf: &ElfFile<'a>) -> Result<(), Error> {
+        let note_header = elf.find_section_by_name(".note.x86_64-fortanix-unknown-sgx").ok_or_else(|| format_err!("Could not find .note.x86_64-fortanix-unknown-sgx header!"))?;
+
+        if note_header.get_type() != Ok(ShType::Note) {
+            bail!("Invalid type {:?} for section: .note.x86_64-fortanix-unknown-sgx", note_header.get_type());
+        }
+
+        match note_header.get_data(&elf) {
+            Ok(SectionData::Note64(header, ptr)) => {
+                if header.name(ptr) != "toolchain-version" {
+                    bail!("Expecting 'toolchain-version' as name for .note.x86_64-fortanix-unknown-sgx");
+                }
+
+                let desc = header.desc(ptr);
+                
+                if desc.len() != 4 {
+                    bail!("Expecting 32 bit 'toolchain-version' in .note.x86_64-fortanix-unknown-sgx");
+                }
+
+                // According to entry.S in rust-lang 'desc - toolchain version number, 32-bit LE'
+                let version : u32 = ((desc[0] as u32) << 0) + ((desc[1] as u32) << 8) + ((desc[2] as u32) << 16) + ((desc[3] as u32) << 24);
```

Comment:
You could write this as `let version = u32::from_le_bytes(desc.try_into().unwrap());`. The unwrap is justified by the check in line 200.

Reply:
Absolutey, definitely use that function instead of manually bitshifting

Reply:
fixed, thanks